src/check.rs

```diff
@@ -61,16 +84,94 @@ pub fn is_available(url: &Url, ctx: &CheckContext) -> Result<(), CheckError> {
+        // attribute of (literal) "#n-m" will not exist, but elements with
+        // `id`s n through m should, this parses the ranged n-m anchor and
+        // checks if elements with `id`s n through m do exist
+        static RUST_LINE_HIGLIGHT_RX: Lazy<Regex> =
+            Lazy::new(|| Regex::new(r#"^(?P<start>[0-9]+)-(?P<end>[0-9]+)$"#).unwrap());
+        match RUST_LINE_HIGLIGHT_RX.captures(fragment) {
+            Some(capture) => match (capture.name("start"), capture.name("end")) {
+                (Some(start_str), Some(end_str)) => {
+                    // NOTE: assumes there are less than 2.pow(32) lines in a source file
+                    let start = start_str.as_str().parse::<i32>().unwrap();
+                    let end = end_str.as_str().parse::<i32>().unwrap();
+                    let missing = (start..=end)
+                        .map(|i| i.to_string())
+                        .filter(|i| !fragments.contains(i))
+                        .collect::<Vec<String>>();
+                    if !missing.is_empty() {
+                        Err(CheckError::AnchorRange(
+                            url.clone(),
+                            fragment.to_string(),
+                            missing,
+                        ))
+                    } else {
+                        Ok(())
+                    }
+                }
+                _ => unreachable!("if the regex matches, it should have capture groups"),
+            },
+            None => Err(CheckError::Anchor(url.clone(), fragment.to_string())),
+        }
+    }
+}
+
 /// Check a URL with the "file" scheme for availability. Returns `false` if it is unavailable.
 fn check_file_url(url: &Url, _ctx: &CheckContext) -> Result<(), CheckError> {
     let path = url.to_file_path().unwrap();
 
-    if path.is_file() || path.join("index.html").is_file() {
-        debug!("Linked file at path {} does exist.", path.display());
-        Ok(())
+    // determine the full path by looking if the path points to a directory,
+    // and if so append `index.html`, this is needed as we'll try to read
+    // the file, so `expanded_path` should point to a file not a directory
+    let expanded_path = if path.is_file() {
+        path
+    } else if path.is_dir() && path.join("index.html").is_file() {
+        path.join("index.html")
     } else {
         debug!("Linked file at path {} does not exist!", path.display());
-        Err(CheckError::File(path))
+        return Err(CheckError::File(path));
+    };
+
+    // the URL might contain a fragment, in that case we need to check if
+    // the fragment exists in the file
+    match url.fragment() {
+        Some(fragment) => check_file_fragment(url, &expanded_path, fragment),
+        None => Ok(()),
+    }
+}
+
+cached_key_result! {
+    CHECK_FILE: SizedCache<(String, String), ()> = SizedCache::with_size(100);
+    Key = { (url.to_string(), fragment.to_string()) };
```

Comment:
Rather than caching whether this exact fragment exists, can we cache the all fragments in the file? Looking up whether a certain fragment exists should be an easy hashmap lookup after that, and it means we get a lot more cache hits without having to parse the file again.